Can't build YandexMobileAds for iOS

Yandex released a new version of their SDK, plus Defold has been updated so I tried building again with Defold 1.9.1 beta and YandexMobileAds 7.1.1

It still doesn't build but the error message has changed, now it complains about iOS SDK version not being at least 13, however it is set to be so in the Podfile:
```
platform :ios, '13.0'
pod 'YandexMobileAds', '7.1.1'
pod 'YandexMobileAdsMediation', '7.1.1'
```

Thee new error:
```
/var/folders/pf/6r_png356xb4lby2dkvym42r0000gn/T/job9582903447976259256/CocoaPodsService/Pods/DivKit/DivKit/Views/DivBlockProvider.swift:77:5: error: concurrency is only available in iOS 13.0.0 or newer
  ) async {
    ^
/var/folders/pf/6r_png356xb4lby2dkvym42r0000gn/T/job9582903447976259256/CocoaPodsService/Pods/DivKit/DivKit/Views/DivBlockProvider.swift:74:8: note: add @available attribute to enclosing instance method
  func setSource(
       ^
```

Full build log
[log.txt.zip](https://github.com/user-attachments/files/16193483/log.txt.zip)

The same result with bob.jar

--- a/server/src/main/java/com/defold/extender/Extender.java
+++ b/server/src/main/java/com/defold/extender/Extender.java
@@ -869,6 +869,16 @@ class Extender {
         List<String> objs = new ArrayList<>();
 
         if (!pod.swiftSourceFiles.isEmpty()) {
+            // Add -enable-experimental-feature AccessLevelOnImport
+            // There are a lot of extensions with Swift code that seem to require this
+            // flag. Perhaps swiftFlags should be added as a new field in ext.manifest?
+            // https://github.com/swiftlang/swift-evolution/blob/main/proposals/0409-access-level-on-imports.md
+            List<String> swiftFlags = (List<String>)mergedContextWithPodsForSwift.get("swiftFlags");
+            List<String> updatedSwiftFlags = new ArrayList<String>(swiftFlags);
+            updatedSwiftFlags.add("-enable-experimental-feature");
+            updatedSwiftFlags.add("AccessLevelOnImport");
+            mergedContextWithPodsForSwift.put("swiftFlags", updatedSwiftFlags);
+
             // generate headers from swift files
             List<String> emitSwiftHeaderCommands = new ArrayList<>();
             LOGGER.info("emit swift header");

--- a/server/src/main/java/com/defold/extender/Extender.java
+++ b/server/src/main/java/com/defold/extender/Extender.java
@@ -813,7 +813,7 @@ class Extender {
     }
 
     // compile the source files of a pod and return a list of object files
-    private List<String> compilePodSourceFiles(PodSpec pod, Map<String, Object> manifestContext) throws IOException, InterruptedException, ExtenderException {
+    private List<String> compilePodSourceFiles(PodSpec pod, Map<String, Object> manifestContext, Map<File, String> sourceToObjectLookup) throws IOException, InterruptedException, ExtenderException {
         // clean up flags from context
         Map<String, Object> trimmedContext = ExtenderUtil.mergeContexts(manifestContext, new HashMap<>());
         trimmedContext.put("flags", new ArrayList<String>());
@@ -882,32 +882,42 @@ class Extender {
             // compile swift source files one by one
             List<String> compileSwiftCommands = new ArrayList<>();
             for (File src : pod.swiftSourceFiles) {
-                final int i = getAndIncreaseNameCount();
-                objs.add(ExtenderUtil.getRelativePath(jobDirectory, o));
+                String objPath = sourceToObjectLookup.get(src);
+                if (objPath == null) {
+                    final int i = getAndIncreaseNameCount();
                     File o = addCompileFileSwift(pod, i, src, mergedContextWithPodsForSwift, compileSwiftCommands);
+                    objPath = ExtenderUtil.getRelativePath(jobDirectory, o);
+                    objs.add(objPath);
+                    sourceToObjectLookup.put(src, objPath);
+                }
             }
             ProcessExecutor.executeCommands(processExecutor, compileSwiftCommands); // in parallel
         }
 
         List<String> commands = new ArrayList<>();
         for (File src : pod.sourceFiles) {
-            String extension = FilenameUtils.getExtension(src.getAbsolutePath());
-            final int i = getAndIncreaseNameCount();
-            File o = null;
-            // use the correct context depending on the source file language
-            if (extension.equals("c")) {
-                o = addCompileFileCppStatic(i, pod.dir, src, mergedContextWithPodsForC, commands);
+            String objPath = sourceToObjectLookup.get(src);
+            if (objPath == null) {
+                String extension = FilenameUtils.getExtension(src.getAbsolutePath());
+                final int i = getAndIncreaseNameCount();
+                File o = null;
+                // use the correct context depending on the source file language
+                if (extension.equals("c")) {
+                    o = addCompileFileCppStatic(i, pod.dir, src, mergedContextWithPodsForC, commands);
+                }
+                else if (extension.equals("m")) {
+                    o = addCompileFileCppStatic(i, pod.dir, src, mergedContextWithPodsForObjC, commands);
+                }
+                else if (extension.equals("mm")) {
+                    o = addCompileFileCppStatic(i, pod.dir, src, mergedContextWithPodsForObjCpp, commands);
+                }
+                else {
+                    o = addCompileFileCppStatic(i, pod.dir, src, mergedContextWithPodsForCpp, commands);
+                }
+                objPath = ExtenderUtil.getRelativePath(jobDirectory, o);
+                objs.add(objPath);
+                sourceToObjectLookup.put(src, objPath);
             }
-            else if (extension.equals("m")) {
-                o = addCompileFileCppStatic(i, pod.dir, src, mergedContextWithPodsForObjC, commands);
-            }
-            else if (extension.equals("mm")) {
-                o = addCompileFileCppStatic(i, pod.dir, src, mergedContextWithPodsForObjCpp, commands);
-            }
-            else {
-                o = addCompileFileCppStatic(i, pod.dir, src, mergedContextWithPodsForCpp, commands);
-            }
-            objs.add(ExtenderUtil.getRelativePath(jobDirectory, o));
         }
         ProcessExecutor.executeCommands(processExecutor, commands); // in parallel
 
@@ -923,6 +933,7 @@ class Extender {
         }
 
         LOGGER.info("buildPods");
+        Map<File, String> sourceToObjectLookup = new HashMap<>();
         for (PodSpec pod : resolvedPods.pods) {
             // The source files of each pod will be compiled and built as a library.
             // We use the same mechanism as when building the extension and create a
@@ -936,7 +947,7 @@ class Extender {
             manifestContext.put("env.IOS_VERSION_MIN", resolvedPods.platformMinVersion);
 
             // Compile pod source files
-            List<String> objs = compilePodSourceFiles(pod, manifestContext);
+            List<String> objs = compilePodSourceFiles(pod, manifestContext, sourceToObjectLookup);
             if (!objs.isEmpty()) {
                 // Create c++ library
                 File lib = createBuildFile(String.format(platformConfig.writeLibPattern, manifestContext.get("extension_name") + "_" + getNameUUID()));

--- a/server/src/main/java/com/defold/extender/Extender.java
+++ b/server/src/main/java/com/defold/extender/Extender.java
@@ -933,6 +933,7 @@ class Extender {
             manifestContext.put("extension_name", pod.name);
             manifestContext.put("extension_name_upper", pod.name.toUpperCase());
             manifestContext.put("osMinVersion", resolvedPods.platformMinVersion);
+            manifestContext.put("env.IOS_VERSION_MIN", resolvedPods.platformMinVersion);
 
             // Compile pod source files
             List<String> objs = compilePodSourceFiles(pod, manifestContext);

--- a/server/src/main/java/com/defold/extender/Extender.java
+++ b/server/src/main/java/com/defold/extender/Extender.java
@@ -561,7 +561,7 @@ class Extender {
     // swiftc: https://gist.github.com/enomoto/7f11d57e4add7e702f9f84f34d3a0f8c
     // swift-frontend: https://gist.github.com/palaniraja/b4de1e64e874b68bda9e5236829cd8a6
 
-    private void emitSwiftHeaders(PodSpec pod, Map<String, Object> manifestContext, List<String> commands) throws IOException, InterruptedException, ExtenderException {
+    private void emitSwiftHeader(PodSpec pod, Map<String, Object> manifestContext, List<String> commands) throws IOException, InterruptedException, ExtenderException {
         List<String> includes = getIncludeDirs(pod.dir);
 
         List<String> frameworks = new ArrayList<>();
@@ -871,7 +871,7 @@ class Extender {
         if (!pod.swiftSourceFiles.isEmpty()) {
             // generate headers from swift files
             List<String> emitSwiftHeaderCommands = new ArrayList<>();
-            emitSwiftHeaders(pod, mergedContextWithPodsForSwift, emitSwiftHeaderCommands);
+            emitSwiftHeader(pod, mergedContextWithPodsForSwift, emitSwiftHeaderCommands);
             ProcessExecutor.executeCommands(processExecutor, emitSwiftHeaderCommands); // in parallel
 
             // generate swift module from swift files

--- a/server/src/main/java/com/defold/extender/Extender.java
+++ b/server/src/main/java/com/defold/extender/Extender.java
@@ -813,7 +813,7 @@ class Extender {
     }
 
     // compile the source files of a pod and return a list of object files
-    private List<String> compilePodSourceFiles(PodSpec pod, Map<String, Object> manifestContext, Map<File, String> sourceToObjectLookup) throws IOException, InterruptedException, ExtenderException {
+    private List<String> compilePodSourceFiles(PodSpec pod, Map<String, Object> manifestContext, Set<File> podSourceLookup) throws IOException, InterruptedException, ExtenderException {
         // clean up flags from context
         Map<String, Object> trimmedContext = ExtenderUtil.mergeContexts(manifestContext, new HashMap<>());
         trimmedContext.put("flags", new ArrayList<String>());
@@ -894,13 +894,12 @@ class Extender {
             // compile swift source files one by one
             List<String> compileSwiftCommands = new ArrayList<>();
             for (File src : pod.swiftSourceFiles) {
-                String objPath = sourceToObjectLookup.get(src);
-                if (objPath == null) {
+                if (!podSourceLookup.contains(src)) {
                     final int i = getAndIncreaseNameCount();
                     File o = addCompileFileSwift(pod, i, src, mergedContextWithPodsForSwift, compileSwiftCommands);
-                    objPath = ExtenderUtil.getRelativePath(jobDirectory, o);
+                    String objPath = ExtenderUtil.getRelativePath(jobDirectory, o);
                     objs.add(objPath);
-                    sourceToObjectLookup.put(src, objPath);
+                    podSourceLookup.add(src);
                 }
             }
             LOGGER.info("compiling {} swift files", compileSwiftCommands.size());
@@ -909,8 +908,7 @@ class Extender {
 
         List<String> commands = new ArrayList<>();
         for (File src : pod.sourceFiles) {
-            String objPath = sourceToObjectLookup.get(src);
-            if (objPath == null) {
+            if (!podSourceLookup.contains(src)) {
                 String extension = FilenameUtils.getExtension(src.getAbsolutePath());
                 final int i = getAndIncreaseNameCount();
                 File o = null;
@@ -927,9 +925,9 @@ class Extender {
                 else {
                     o = addCompileFileCppStatic(i, pod.dir, src, mergedContextWithPodsForCpp, commands);
                 }
-                objPath = ExtenderUtil.getRelativePath(jobDirectory, o);
+                String objPath = ExtenderUtil.getRelativePath(jobDirectory, o);
                 objs.add(objPath);
-                sourceToObjectLookup.put(src, objPath);
+                podSourceLookup.add(src);
             }
         }
         LOGGER.info("compiling {} source files", commands.size());
@@ -947,7 +945,7 @@ class Extender {
         }
 
         LOGGER.info("buildPods");
-        Map<File, String> sourceToObjectLookup = new HashMap<>();
+        Set<File> podSourceLookup = new HashSet<>();
         for (PodSpec pod : resolvedPods.pods) {
             LOGGER.info("building {}", pod.name);
             // The source files of each pod will be compiled and built as a library.
@@ -962,7 +960,7 @@ class Extender {
             manifestContext.put("env.IOS_VERSION_MIN", resolvedPods.platformMinVersion);
 
             // Compile pod source files
-            List<String> objs = compilePodSourceFiles(pod, manifestContext, sourceToObjectLookup);
+            List<String> objs = compilePodSourceFiles(pod, manifestContext, podSourceLookup);
             if (!objs.isEmpty()) {
                 // Create c++ library
                 File lib = createBuildFile(String.format(platformConfig.writeLibPattern, manifestContext.get("extension_name") + "_" + getNameUUID()));

--- a/server/src/main/java/com/defold/extender/Extender.java
+++ b/server/src/main/java/com/defold/extender/Extender.java
@@ -871,20 +871,20 @@ class Extender {
         if (!pod.swiftSourceFiles.isEmpty()) {
             // generate headers from swift files
             List<String> emitSwiftHeaderCommands = new ArrayList<>();
-            emitSwiftHeaders(pod, mergedContextWithPodsForC, emitSwiftHeaderCommands);
+            emitSwiftHeaders(pod, mergedContextWithPodsForSwift, emitSwiftHeaderCommands);
             ProcessExecutor.executeCommands(processExecutor, emitSwiftHeaderCommands); // in parallel
 
             // generate swift module from swift files
             List<String> emitSwiftModuleCommands = new ArrayList<>();
-            emitSwiftModule(pod, mergedContextWithPodsForC, emitSwiftModuleCommands);
+            emitSwiftModule(pod, mergedContextWithPodsForSwift, emitSwiftModuleCommands);
             ProcessExecutor.executeCommands(processExecutor, emitSwiftModuleCommands); // in parallel
 
             // compile swift source files one by one
             List<String> compileSwiftCommands = new ArrayList<>();
             for (File src : pod.swiftSourceFiles) {
                 final int i = getAndIncreaseNameCount();
-                File o = addCompileFileSwift(pod, i, src, mergedContextWithPodsForC, compileSwiftCommands);
                 objs.add(ExtenderUtil.getRelativePath(jobDirectory, o));
+                    File o = addCompileFileSwift(pod, i, src, mergedContextWithPodsForSwift, compileSwiftCommands);
             }
             ProcessExecutor.executeCommands(processExecutor, compileSwiftCommands); // in parallel
         }

--- a/server/src/main/java/com/defold/extender/Extender.java
+++ b/server/src/main/java/com/defold/extender/Extender.java
@@ -871,11 +871,13 @@ class Extender {
         if (!pod.swiftSourceFiles.isEmpty()) {
             // generate headers from swift files
             List<String> emitSwiftHeaderCommands = new ArrayList<>();
+            LOGGER.info("emit swift header");
             emitSwiftHeader(pod, mergedContextWithPodsForSwift, emitSwiftHeaderCommands);
             ProcessExecutor.executeCommands(processExecutor, emitSwiftHeaderCommands); // in parallel
 
             // generate swift module from swift files
             List<String> emitSwiftModuleCommands = new ArrayList<>();
+            LOGGER.info("emit swift module");
             emitSwiftModule(pod, mergedContextWithPodsForSwift, emitSwiftModuleCommands);
             ProcessExecutor.executeCommands(processExecutor, emitSwiftModuleCommands); // in parallel
 
@@ -891,6 +893,7 @@ class Extender {
                     sourceToObjectLookup.put(src, objPath);
                 }
             }
+            LOGGER.info("compiling {} swift files", compileSwiftCommands.size());
             ProcessExecutor.executeCommands(processExecutor, compileSwiftCommands); // in parallel
         }
 
@@ -919,6 +922,7 @@ class Extender {
                 sourceToObjectLookup.put(src, objPath);
             }
         }
+        LOGGER.info("compiling {} source files", commands.size());
         ProcessExecutor.executeCommands(processExecutor, commands); // in parallel
 
         return objs;
@@ -935,6 +939,7 @@ class Extender {
         LOGGER.info("buildPods");
         Map<File, String> sourceToObjectLookup = new HashMap<>();
         for (PodSpec pod : resolvedPods.pods) {
+            LOGGER.info("building {}", pod.name);
             // The source files of each pod will be compiled and built as a library.
             // We use the same mechanism as when building the extension and create a
             // manifest context for each pod
@@ -954,6 +959,7 @@ class Extender {
                 Map<String, Object> context = createContext(manifestContext);
                 context.put("tgt", lib);
                 context.put("objs", objs);
+                LOGGER.info("creating library {} from {} objects", lib.getName(), objs.size());
                 executeCommand(platformConfig.libCmd, context);
             }
         }

--- a/server/src/main/java/com/defold/extender/Extender.java
+++ b/server/src/main/java/com/defold/extender/Extender.java
@@ -2213,6 +2213,7 @@ class Extender {
                 podAppContext.put("libs", resolvedPods.getAllPodLibs(platform));
                 podAppContext.put("linkFlags", resolvedPods.getAllPodLinkFlags(platform));
                 podAppContext.put("osMinVersion", resolvedPods.platformMinVersion);
+                podAppContext.put("env.IOS_VERSION_MIN", resolvedPods.platformMinVersion);
             }
             Map mergedAppContextWithPods = ExtenderUtil.mergeContexts(mergedAppContext, podAppContext);
 

--- a/server/src/main/java/com/defold/extender/Extender.java
+++ b/server/src/main/java/com/defold/extender/Extender.java
@@ -862,7 +862,9 @@ class Extender {
         mergedContextWithPodsForObjC.put("systemIncludes", new ArrayList<String>());
         mergedContextWithPodsForObjCpp.put("systemIncludes", new ArrayList<String>());
         mergedContextWithPodsForSwift.put("systemIncludes", new ArrayList<String>());
-
+        // also clearing systemIncludes for c
+        // see https://github.com/defold/extender/issues/389
+        mergedContextWithPodsForC.put("systemIncludes", new ArrayList<String>());
 
         List<String> objs = new ArrayList<>();
 

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -1314,19 +1314,19 @@ public class CocoaPodsService {
 
         // find all podfiles and filter down to a list of podfiles specifically
         // for the platform we are resolving pods for
-        List<File> allPodFiles = ExtenderUtil.listFilesMatchingRecursive(jobDir, "Podfile");
-        List<File> platformPodFiles = new ArrayList<>();
-        for (File podFile : allPodFiles) {
+        List<File> allPodfiles = ExtenderUtil.listFilesMatchingRecursive(jobDir, "Podfile");
+        List<File> platformPodfiles = new ArrayList<>();
+        for (File podFile : allPodfiles) {
             String parentFolder = podFile.getParentFile().getName();
             if ((platform.contains("ios") && parentFolder.contains("ios")) ||
                 (platform.contains("osx") && parentFolder.contains("osx"))) {
-                platformPodFiles.add(podFile);
+                platformPodfiles.add(podFile);
             }
             else {
                 LOGGER.warn("Unexpected Podfile found in " + podFile);
             }
         }
-        if (platformPodFiles.isEmpty()) {
+        if (platformPodfiles.isEmpty()) {
             LOGGER.info("Project has no Cocoapod dependencies");
             return null;
         }

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -297,7 +297,7 @@ public class CocoaPodsService {
                     return spec;
                 }
             }
-            return  null;
+            return null;
         }
 
         @Override

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -31,6 +31,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.HashMap;
 import java.util.Set;
+import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.Iterator;
 import java.util.Collections;
@@ -436,7 +437,7 @@ public class CocoaPodsService {
                         mainPodfile.platformMinVersion = version;
                     }
                 }
-                else {
+                else if (line.length() > 1) {
                     pods.add(line);
 
                     // get the pod name from the line
@@ -879,13 +880,10 @@ public class CocoaPodsService {
             headers.add(swiftModuleHeader.getAbsolutePath());
         }
 
-        // module name, including parent spec module name
-        String fullModuleName = ((pod.parentSpec != null) ? (pod.parentSpec.moduleName + "_") : "") + pod.moduleName;
-
         // generate final modulemap
         HashMap<String, Object> envContext = new HashMap<>();
         envContext.put("FRAMEWORKOPT", pod.containsFramework ? "framework" : "");
-        envContext.put("MODULE_ID", fullModuleName);
+        envContext.put("MODULE_ID", pod.moduleName);
         envContext.put("HEADERS", headers);
         envContext.put("UMBRELLA_HEADER", umbrellaHeader);
         envContext.put("SUBMODULE", pod.containsFramework ? "module * { export * }" : "");
@@ -910,17 +908,43 @@ public class CocoaPodsService {
         return moduleMapFile.getAbsolutePath();
     }
 
+
+    private String toC99extIdentifier(String s) {
+        return s.replaceAll("^([0-9])", "_$1")  // 123FooBar -> _123FooBar
+            .replaceAll("[\\+].*", "")          // NSData+zlib -> NSData
+            .replaceAll("[^a-zA-Z0-9_]", "_")   // Foo-Bar -> Foo_Bar
+            .replaceAll("_+", "_");             // Foo__Bar -> Foo_Bar
+    }
+
+    // It seems like some exceptions such as KSCrash use a module name
+    // which doesn't separate parent and subspec by an underscore
+    // https://github.com/kstenerud/KSCrash/blob/master/KSCrash.podspec#L21
+    private static Set<String> NAME_EXCEPTIONS = new HashSet<>(Arrays.asList("KSCrash"));
+
+    // https://github.com/CocoaPods/Core/blob/master/lib/cocoapods-core/specification.rb#L187
+    private String getModuleName(JSONObject specJson, PodSpec parent) {
+        if (specJson.containsKey("module_name")) {
+            return (String)specJson.get("module_name");
+        }
+
+        if (specJson.containsKey("header_dir")) {
+            return toC99extIdentifier((String)specJson.get("header_dir"));
+        }
+
+        String name = (String)specJson.get("name");
+        String fixedName = toC99extIdentifier(name);
+        if (parent == null) {
+            return fixedName;
+        }
+
+        return parent.name + (NAME_EXCEPTIONS.contains(parent.name) ? "" : "_") + fixedName;
+    }
+
     // https://guides.cocoapods.org/syntax/podspec.html
     private PodSpec createPodSpec(JSONObject specJson, PodSpec parent, File podsDir, File generatedDir, File jobDir, Map<String, Object> jobEnvContext) throws ExtenderException {
         PodSpec spec = new PodSpec();
         spec.name = (String)specJson.get("name");
-        if (specJson.containsKey("module_name")) {
-            spec.moduleName = (String)specJson.get("module_name");
-        }
-        else {
-            // NSData+zlib -> NSData
-            spec.moduleName = ((String)specJson.get("name")).replaceAll("[\\+].*", "");
-        }
+        spec.moduleName = getModuleName(specJson, parent);
         spec.version = (parent == null) ? (String)specJson.get("version") : parent.version;
         spec.dir = (parent == null) ? new File(podsDir, spec.name) : parent.dir;
         spec.parentSpec = parent;
@@ -938,6 +962,7 @@ public class CocoaPodsService {
             spec.defines.osx.addAll(parent.defines.osx);
             spec.linkflags.ios.addAll(parent.linkflags.ios);
             spec.linkflags.ios.addAll(parent.linkflags.ios);
+            spec.flags.remove("-fmodule-name=" + parent.moduleName);
         }
 
         // platform versions
@@ -997,11 +1022,12 @@ public class CocoaPodsService {
         spec.flags.osx.objc.add("-fmodules");
         spec.flags.ios.objcpp.add("-fcxx-modules");
         spec.flags.osx.objcpp.add("-fcxx-modules");
-        if (spec.parentSpec == null) {
-            spec.flags.ios.objc.add("-fmodule-name=" + spec.moduleName);
-            spec.flags.osx.objc.add("-fmodule-name=" + spec.moduleName);
-            spec.flags.ios.objcpp.add("-fmodule-name=" + spec.moduleName);
-        }
+        spec.flags.ios.c.add("-fmodule-name=" + spec.moduleName);
+        spec.flags.osx.c.add("-fmodule-name=" + spec.moduleName);
+        spec.flags.ios.objc.add("-fmodule-name=" + spec.moduleName);
+        spec.flags.osx.objc.add("-fmodule-name=" + spec.moduleName);
+        spec.flags.ios.objcpp.add("-fmodule-name=" + spec.moduleName);
+        spec.flags.osx.objcpp.add("-fmodule-name=" + spec.moduleName);
 
         // resources
         // https://guides.cocoapods.org/syntax/podspec.html#resources
@@ -1369,6 +1395,10 @@ public class CocoaPodsService {
         resolvedPods.podsDir = new File(workingDir, "Pods");
         resolvedPods.frameworksDir = frameworksDir;
         resolvedPods.generatedDir = generatedDir;
+
+        LOGGER.info("Resolved Cocoapod dependencies");
+        LOGGER.info(resolvedPods.toString());
+
         return resolvedPods;
     }
 

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -925,13 +925,6 @@ public class CocoaPodsService {
         spec.dir = (parent == null) ? new File(podsDir, spec.name) : parent.dir;
         spec.parentSpec = parent;
 
-        if (parent == null) {
-            spec.installed = spec.dir.exists();
-        }
-        else {
-            spec.installed = (new File(new File(spec.dir, spec.parentSpec.name), spec.name)).exists();
-        }
-
         // generated files relating to the pod
         // modulemap, swift header etc
         spec.generatedDir = new File(generatedDir, spec.name);
@@ -1102,6 +1095,7 @@ public class CocoaPodsService {
                 }
             }
         }
+        spec.installed = spec.containsFramework || !spec.sourceFiles.isEmpty() || !spec.swiftSourceFiles.isEmpty();
 
         // add swift libs to the runtime search path
         if (!spec.swiftSourceFiles.isEmpty()) {

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -985,6 +985,8 @@ public class CocoaPodsService {
         spec.flags.osx.objcpp.add("--language=objective-c++");
         // CocoaPods sets CLANG_ENABLE_MODULES when creating an XCode project
         // https://xcodebuildsettings.com/#clang_enable_modules
+        spec.flags.ios.c.add("-fmodules");
+        spec.flags.osx.c.add("-fmodules");
         spec.flags.ios.objc.add("-fmodules");
         spec.flags.osx.objc.add("-fmodules");
         spec.flags.ios.objcpp.add("-fcxx-modules");

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -1291,6 +1291,9 @@ public class CocoaPodsService {
         Collections.reverse(reversePodnames);
         installedPods.pods.addAll(getSpecsAndDependencies(installedPods.podsMap, reversePodnames));
         LOGGER.info("Installed pods");
+        for (PodSpec pod : installedPods.pods) {
+            LOGGER.info("  " + pod.name);
+        }
         return installedPods;
     }
 

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -176,6 +176,13 @@ public class CocoaPodsService {
             objcpp.addAll(set.objcpp);
         }
 
+        public void remove(String value) {
+            c.remove(value);
+            cpp.remove(value);
+            objc.remove(value);
+            objcpp.remove(value);
+        }
+
         @Override
         public String toString() {
             StringBuilder sb = new StringBuilder();
@@ -200,6 +207,11 @@ public class CocoaPodsService {
                 ios.add(v);
                 osx.add(v);
             }
+        }
+
+        public void remove(String value) {
+            ios.remove(value);
+            osx.remove(value);
         }
 
         @Override

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -40,6 +40,24 @@ import java.lang.StringBuilder;
 @Service
 public class CocoaPodsService {
 
+    private class MainPodfile {
+        public List<String> podnames = new ArrayList<>();
+        public String platformMinVersion;
+        public String platform;
+        public File file;
+
+        @Override
+        public String toString() {
+            StringBuilder sb = new StringBuilder();
+            sb.append("file: " + file);
+            sb.append("pod count: " + podnames.size() + "\n");
+            sb.append("pod names: " + podnames + "\n");
+            sb.append("platform: " + platform + "\n");
+            sb.append("min version: " + platformMinVersion + "\n");
+            return sb.toString();
+        }
+    }
+
     public class ResolvedPods {
         public List<PodSpec> pods = new ArrayList<>();
         public File podsDir;
@@ -367,18 +385,20 @@ public class CocoaPodsService {
      * @param jobDirectory The job directory from where to search for Podfiles
      * @param workingDir The working directory where pods should be resolved
      * @param platform For which platform to resolve pods
-     * @return Minimum platform version
+     * @return Main pod file structure
      */
-    private String createMainPodFile(List<File> podFiles, File jobDirectory, File workingDir, String platform, Map<String, Object> jobEnvContext) throws IOException {
-        File mainPodFile = new File(workingDir, "Podfile");
+    private MainPodfile createMainPodfile(List<File> podFiles, File jobDirectory, File workingDir, String platform, Map<String, Object> jobEnvContext) throws IOException {
+        MainPodfile mainPodfile = new MainPodfile();
+
+        mainPodfile.file = new File(workingDir, "Podfile");
 
         // This file might exist when testing and debugging the extender using a debug job folder
-        podFiles.remove(mainPodFile);
+        podFiles.remove(mainPodfile.file);
 
-        String mainPodfilePlatformVersion = (platform.contains("ios") ? 
+        mainPodfile.platformMinVersion = (platform.contains("ios") ? 
             jobEnvContext.get("env.IOS_VERSION_MIN").toString(): 
             jobEnvContext.get("env.MACOS_VERSION_MIN").toString());
-        String mainPodfilePlatform = (platform.contains("ios") ? "ios" : "osx");
+        mainPodfile.platform = (platform.contains("ios") ? "ios" : "osx");
 
         // Load all Podfiles
         List<String> pods = new ArrayList<>();
@@ -394,27 +414,33 @@ public class CocoaPodsService {
             for (String line : lines) {
                 if (line.startsWith("platform :")) {
                     String version = line.replaceFirst("platform :ios|platform :osx", "").replace(",", "").replace("'", "").trim();
-                    if (!version.isEmpty() && (compareVersions(version, mainPodfilePlatformVersion) > 0)) {
-                        mainPodfilePlatformVersion = version;
+                    if (!version.isEmpty() && (compareVersions(version, mainPodfile.platformMinVersion) > 0)) {
+                        mainPodfile.platformMinVersion = version;
                     }
                 }
                 else {
                     pods.add(line);
+
+                    // get the pod name from the line
+                    // example: pod 'KSCrash', '1.17.4' -> KSCrash
+                    String[] parts = line.split("'");
+                    String podname = parts[1];
+                    mainPodfile.podnames.add(podname);
                 }
             }
         }
 
         // Create main Podfile contents
         HashMap<String, Object> envContext = new HashMap<>();
-        envContext.put("PLATFORM", mainPodfilePlatform);
-        envContext.put("PLATFORM_VERSION", mainPodfilePlatformVersion);
+        envContext.put("PLATFORM", mainPodfile.platform);
+        envContext.put("PLATFORM_VERSION", mainPodfile.platformMinVersion);
         envContext.put("PODS", pods);
-        String mainPodFileContents = templateExecutor.execute(podfileTemplateContents, envContext);
-        LOGGER.info("Created main Podfile:\n{}", mainPodFileContents);
+        String mainPodfileContents = templateExecutor.execute(podfileTemplateContents, envContext);
+        LOGGER.info("Created main Podfile:\n{}", mainPodfileContents);
 
-        Files.write(mainPodFile.toPath(), mainPodFileContents.getBytes());
+        Files.write(mainPodfile.file.toPath(), mainPodfileContents.getBytes());
 
-        return mainPodfilePlatformVersion;
+        return mainPodfile;
     }
 
     private boolean isHeaderFile(String filename) {
@@ -1309,8 +1335,8 @@ public class CocoaPodsService {
         workingDir.mkdirs();
         generatedDir.mkdirs();
 
-        String platformMinVersion = createMainPodFile(platformPodFiles, jobDir, workingDir, platform, jobEnvContext);
         List<PodSpec> pods = installPods(jobDir, workingDir, generatedDir, jobEnvContext);
+        MainPodfile mainPodfile = createMainPodfile(platformPodfiles, jobDir, workingDir, platform, jobEnvContext);
         copyPodFrameworks(pods, frameworksDir, config);
 
         dumpDir(jobDir, 0);
@@ -1319,7 +1345,7 @@ public class CocoaPodsService {
 
         ResolvedPods resolvedPods = new ResolvedPods();
         resolvedPods.pods = pods;
-        resolvedPods.platformMinVersion = platformMinVersion;
+        resolvedPods.platformMinVersion = mainPodfile.platformMinVersion;
         resolvedPods.podsDir = new File(workingDir, "Pods");
         resolvedPods.frameworksDir = frameworksDir;
         resolvedPods.generatedDir = generatedDir;

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -417,6 +417,10 @@ public class CocoaPodsService {
         return mainPodfilePlatformVersion;
     }
 
+    private boolean isHeaderFile(String filename) {
+        return filename.endsWith(".h") || filename.endsWith(".def");
+    }
+
     /**
      * Add source files matching a pattern to a pod
      * @param pod
@@ -431,7 +435,7 @@ public class CocoaPodsService {
                 pod.swiftSourceFiles.add(podSrcFile);
                 pod.swiftSourceFilePaths.add(podSrcFile.getAbsolutePath());
             }
-            else if (!filename.endsWith(".h")) {
+            else if (!isHeaderFile(filename)) {
                 pod.sourceFiles.add(podSrcFile);
             }
         }
@@ -446,7 +450,8 @@ public class CocoaPodsService {
         String absolutePatternPath = pod.dir.getAbsolutePath() + File.separator + pattern;
         List<File> podSrcFiles = listFilesGlob(pod.dir, absolutePatternPath);
         for (File podSrcFile : podSrcFiles) {
-            if (podSrcFile.getName().endsWith(".h")) {
+            final String filename = podSrcFile.getName();
+            if (isHeaderFile(filename)) {
                 File podIncludeDir = podSrcFile.getParentFile();
                 if (podIncludeDir != null) {
                     pod.includePaths.add(podIncludeDir);

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -284,7 +284,6 @@ public class CocoaPodsService {
         public Set<String> swiftSourceFilePaths = new LinkedHashSet<>();
         public Set<File> swiftSourceFiles = new LinkedHashSet<>();
         public Set<File> sourceFiles = new LinkedHashSet<>();
-        public String sourceFilePatterns = "";
         public Set<File> includePaths = new LinkedHashSet<>();
         public PodSpec parentSpec = null;
         public List<String> defaultSubspecs = new ArrayList<>();
@@ -1106,7 +1105,6 @@ public class CocoaPodsService {
         // https://guides.cocoapods.org/syntax/podspec.html#source_files
         JSONArray sourceFiles = getAsJSONArray(specJson, "source_files");
         if (sourceFiles != null) {
-            spec.sourceFilePatterns = sourceFiles.toString();
             Iterator<String> it = sourceFiles.iterator();
             while (it.hasNext()) {
                 String path = it.next();

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -1072,7 +1072,6 @@ public class CocoaPodsService {
 
         // find source and header files
         // https://guides.cocoapods.org/syntax/podspec.html#source_files
-        List<String> sourceFilePatterns = new ArrayList<>();
         JSONArray sourceFiles = getAsJSONArray(specJson, "source_files");
         if (sourceFiles != null) {
             Iterator<String> it = sourceFiles.iterator();

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -58,6 +58,11 @@ public class CocoaPodsService {
         }
     }
 
+    private class InstalledPods {
+        public Map<String, PodSpec> podsMap = new HashMap<>();
+        public List<PodSpec> pods = new ArrayList<>();
+    }
+
     public class ResolvedPods {
         public List<PodSpec> pods = new ArrayList<>();
         public File podsDir;
@@ -1207,10 +1212,12 @@ public class CocoaPodsService {
      * @param jobDir Directory of entire build job
      * @param workingDir Directory where to install pods. The directory must contain a valid Podfile
      * @param jobEnvContext Job environment context which contains all the job environment variables with `env.*` keys
-     * @return A list of PodSpec instances for the installed pods
+     * @return An InstalledPods object with installed pods
      */
-    private List<PodSpec> installPods(File jobDir, File workingDir, File generatedDir, Map<String, Object> jobEnvContext) throws IOException, ExtenderException {
+    private InstalledPods installPods(File jobDir, File workingDir, File generatedDir, Map<String, Object> jobEnvContext) throws IOException, ExtenderException {
         LOGGER.info("Installing pods");
+        InstalledPods installedPods = new InstalledPods();
+
         File podFile = new File(workingDir, "Podfile");
         if (!podFile.exists()) {
             throw new ExtenderException("Unable to find Podfile " + podFile);
@@ -1252,7 +1259,6 @@ public class CocoaPodsService {
 
         // get all the pod specs and store them in a map, keyed on pod name
         File podsDir = new File(workingDir, "Pods");
-        Map<String, PodSpec> specsMap = new HashMap<>();
         for (String podname : podnames) {
             // 'GoogleUtilities/Environment (7.10.0)'  -> 'GoogleUtilities/Environment' -> ['GoogleUtilities', 'Environment']
             String podnameparts[] = splitPodname(podname);
@@ -1260,7 +1266,7 @@ public class CocoaPodsService {
             String mainpodname = podnameparts[0];
             // 'GoogleUtilities/Environment (7.10.0)'  -> '7.10.0'
             String podversion = podname.replaceFirst(".*\\(", "").replace(")", "");
-            if (!specsMap.containsKey(mainpodname)) {
+            if (!installedPods.podsMap.containsKey(mainpodname)) {
                 String cmd = "pod spec cat --regex ^" + mainpodname + "$ --version=" + podversion;
                 String specJson = execCommand(cmd).replace(cmd, "");
                 // find first occurence of { because in some cases pod command
@@ -1273,17 +1279,16 @@ public class CocoaPodsService {
                 //     "dependencies": {
                 //     "GoogleAppMeasurement": [
                 specJson = specJson.substring(specJson.indexOf("{", 0), specJson.length());
-
-                specsMap.put(mainpodname, createPodSpec(specJson, podsDir, generatedDir, jobDir, jobEnvContext));
+                installedPods.podsMap.put(mainpodname, createPodSpec(specJson, podsDir, generatedDir, jobDir, jobEnvContext));
             }
         }
 
         // build list of specs and their dependencies (dependencies first)
         List<String> reversePodnames = new ArrayList<>(podnames);
         Collections.reverse(reversePodnames);
-        List<PodSpec> specs = new ArrayList<PodSpec>(getSpecsAndDependencies(specsMap, reversePodnames));
+        installedPods.pods.addAll(getSpecsAndDependencies(installedPods.podsMap, reversePodnames));
         LOGGER.info("Installed pods");
-        return specs;
+        return installedPods;
     }
 
     private Map<String, Object> createJobEnvContext(Map<String, Object> env) {
@@ -1335,8 +1340,11 @@ public class CocoaPodsService {
         workingDir.mkdirs();
         generatedDir.mkdirs();
 
-        List<PodSpec> pods = installPods(jobDir, workingDir, generatedDir, jobEnvContext);
         MainPodfile mainPodfile = createMainPodfile(platformPodfiles, jobDir, workingDir, platform, jobEnvContext);
+        InstalledPods installedPods = installPods(jobDir, workingDir, generatedDir, jobEnvContext);
+        List<PodSpec> allPods = installedPods.pods;
+        List<PodSpec> pods = new ArrayList<>();
+        pods = allPods;
         copyPodFrameworks(pods, frameworksDir, config);
 
         dumpDir(jobDir, 0);

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -271,6 +271,7 @@ public class CocoaPodsService {
         public Set<String> swiftSourceFilePaths = new LinkedHashSet<>();
         public Set<File> swiftSourceFiles = new LinkedHashSet<>();
         public Set<File> sourceFiles = new LinkedHashSet<>();
+        public String sourceFilePatterns = "";
         public Set<File> includePaths = new LinkedHashSet<>();
         public PodSpec parentSpec = null;
         public List<String> defaultSubspecs = new ArrayList<>();
@@ -1074,6 +1075,7 @@ public class CocoaPodsService {
         // https://guides.cocoapods.org/syntax/podspec.html#source_files
         JSONArray sourceFiles = getAsJSONArray(specJson, "source_files");
         if (sourceFiles != null) {
+            spec.sourceFilePatterns = sourceFiles.toString();
             Iterator<String> it = sourceFiles.iterator();
             while (it.hasNext()) {
                 String path = it.next();

--- a/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
+++ b/server/src/main/java/com/defold/extender/services/CocoaPodsService.java
@@ -115,6 +115,17 @@ public class CocoaPodsService {
             }
             return new ArrayList<String>(weakFrameworks);
         }
+
+        @Override
+        public String toString() {
+            StringBuilder sb = new StringBuilder();
+            sb.append("pod count: " + pods.size() + "\n");
+            sb.append("pods dir: " + podsDir + "\n");
+            sb.append("frameworks Dir: " + frameworksDir + "\n");
+            sb.append("generated dir: " + generatedDir + "\n");
+            sb.append("platform min version: " + platformMinVersion);
+            return sb.toString();
+        }
     }
 
     public class LanguageSet {

